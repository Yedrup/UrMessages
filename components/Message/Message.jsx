import React from 'react';
import Link from 'next/link';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { useRouter } from 'next/router';
import Avatar, { avatarSize } from '../Avatar';
import getRandomAvatarPhoto from '../../lib/utils/getRandomAvatar';

const StyledMessageItem = styled.article`
  --padding: 1rem 2.5rem;
  padding: 0;
  background: ${({ theme }) => theme.backgroundCard};
  border-radius: ${({ theme }) => theme.borderRadius};
  box-shadow: ${({ theme }) => theme.elevation1};
  &:hover {
    box-shadow: ${({ theme }) => theme.elevation8};
  }
  header {
    display: flex;
    flex-direction: column;
    align-items: flex-start;
    padding: var(--padding);
    color: #aaeeee;
    & h3 {
      display: flex;
      justify-content: space-between;
      width: 100%;
    }
    .user {
      margin-right: 1rem;
      display: flex;
      span {
        margin-left: 1rem;
      }
      .username {
        color: #aaeeeede;
      }
    }
  }
  section {
    padding: var(--padding);
    overflow-wrap: anywhere;
    text-align: justify;
    color: ${({ theme }) => theme.offWhite};
    p {
      background: #0000002b;
      padding: 2rem 1rem;
    }
  }
  footer {
    padding: var(--padding);
    text-align: right;
    color: grey;
    p {
      margin: 0.5rem;
    }
    .small-text {
      font-size: 1.3rem;
    }
  }
`;

const Message = ({ message }) => {
  if (!message) return null;
  const { title, isPublic, content, id, isThreadAllowed, date, user } = message;
  const router = useRouter();

  // fallback in case there is no router.query.type
  const type = router?.query?.type
    ? router?.query?.type
    : isPublic
    ? 'public'
    : 'private';

  const imageUrl = getRandomAvatarPhoto(avatarSize);
  // check if the current message is displayed on a thread page
  const areWeOnMessageThreadPage = !!router?.query?.id;

  return (
    <StyledMessageItem data-testid="message-item">
      <header>
        <h3>
          <span>{title}</span>
          <span>{isPublic ? 'ðŸ“£' : 'ðŸ”’'} </span>
        </h3>
        <div className="user">
          <Avatar imageUrl={imageUrl} />
          <span className="username">{user}</span>
        </div>
      </header>
      <section>
        <p>{content}</p>
      </section>
      <footer>
        <p>{date}</p>
        {/* <p className="small-text">id:{id}</p> */}
        {isThreadAllowed && !areWeOnMessageThreadPage && (
          <Link
            href={{
              pathname: '/messages/[type]/[id]',
              query: { type, isPublic },
            }}
            as={`/messages/${type}/${id}`}
            key={id}
          >
            <a data-testid="thread-link" className="c-button--text">
              SEE THREAD â†’
            </a>
          </Link>
        )}
      </footer>
    </StyledMessageItem>
  );
};

export default Message;

Message.propTypes = {
  message: PropTypes.object,
};
