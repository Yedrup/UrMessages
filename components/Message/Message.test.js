import { render, cleanup } from '@testing-library/react';

import Message from './Message';

import {
  fakePrivateMessage,
  fakePublicMessage,
  fakeMessageNoThreadAllowed,
} from '../../lib/testUtils';

describe('<Message/>', () => {
  afterEach(cleanup);

  it('properly renders the component - snapshot', () => {
    const { asFragment } = render(<Message message={fakePrivateMessage} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('It should not have a thread link cause the threadAllowed is false', () => {
    const { queryByTestId } = render(
      <Message message={fakeMessageNoThreadAllowed} />
    );
    expect(queryByTestId('thread-link')).not.toBeInTheDocument();
  });

  it('It should display a thread link cause threadAllowed is true', () => {
    const { queryByTestId } = render(<Message message={fakePublicMessage} />);
    expect(queryByTestId('thread-link')).toBeInTheDocument();
  });
});
