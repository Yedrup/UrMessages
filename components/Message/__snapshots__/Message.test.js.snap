// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Message/> properly renders the component - snapshot 1`] = `
<DocumentFragment>
  <article
    class="Message__StyledMessageItem-uraoec-0 dDlRBV"
    data-testid="message-item"
  >
    <header>
      <h3>
        Molestias molestias 
        <span>
          ðŸ”’
        </span>
      </h3>
    </header>
    <section>
      reiciendis asperiores tempore enim earum velit consequatur
      <p>
        e0f17993-697b-4807-88b9-93e139396ddc
      </p>
    </section>
    <footer>
      <p>
        Sunday, August 2, 2020, 1:02:01 PM
      </p>
      <a
        class="c-button--text"
        data-testid="thread-link"
        href="/messages/private/e0f17993-697b-4807-88b9-93e139396ddc"
      >
        SEE THREAD â†’
      </a>
    </footer>
  </article>
</DocumentFragment>
`;
